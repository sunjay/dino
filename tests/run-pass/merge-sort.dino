//! A dino program that reads a line of text as input, sorts it, and then outputs the
//! sorted string.

fn main() {
    // Continue forever until the user quits the program with Ctrl-C
    while true {
        print_bstr(b"Enter some text: ");
        let input = read_line();
        let output = merge_sort(input);
        print_bstr(output);
    }
}

fn merge_sort(input: bstr) -> bstr {
    let input_len = bstr_len(input);

    // Base case: return a single character (or nothing)
    if int_lte(input_len, 1) {
        return input;
    }

    // Split the string and sort each half
    let left = bstr_slice(input, 0, int_div(input_len, 2));
    let right = bstr_slice(input, int_div(input_len, 2), input_len);

    let sorted_left = merge_sort(left);
    let sorted_right = merge_sort(right);

    // Merge the two halves together to get a single sorted string
    merge(sorted_left, sorted_right)
}

/// Merges two sorted strings into a single sorted string
fn merge(left: bstr, right: bstr) -> bstr {
    let left_len = bstr_len(left);
    let right_len = bstr_len(right);

    let left_i = 0;
    let right_i = 0;

    let output = b"";
    // Continue until one of the strings runs out of characters
    while bool_and(int_lt(left_i, left_len), int_lt(right_i, right_len)) {
        let left_byte = bstr_get(left, left_i);
        let right_byte = bstr_get(right, right_i);

        if byte_lt(left_byte, right_byte) {
            append_string_byte(output, left_byte);
            left_i = add_int(left_i, 1);
        } else { // byte_gte(left_byte, right_byte)
            append_string_byte(output, right_byte);
            right_i = add_int(right_i, 1);
        }
    }

    // Append any remaining characters on to the string
    // Only one of these loops will run
    while int_lt(left_i, left_len) {
        append_string_byte(output, bstr_get(left, left_i));
        left_i = add_int(left_i, 1);
    }
    while int_lt(right_i, right_len) {
        append_string_byte(output, bstr_get(right, right_i));
        right_i = add_int(right_i, 1);
    }

    output
}
