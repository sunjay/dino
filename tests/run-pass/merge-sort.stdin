abcd
1234
dcba
dbac

1
111
8959404457
d4WBa$$6AL82bc*#Z^4a^P5DHZXC(6)PZFG)
9O0af%1CQ6UXVKW2fa48KZcQ7cAN71P$0BSTWc@^J*5A1C*H)4PKRQaFFTN(P^D9SCabKMG1&Gf^PJ!SHcT#DKb8#!Ec&H2dL&dM2QZN6!3bMN(7N7^^S)#UKN@H4@2D
MPE3cWO3d7VLWeSbVbaXD
PFALVWZYWT
bmwwafyncmpqzrlmesqpmpdylpirjsmokqdpiszwkgasoqydgchklmzddlmgit
