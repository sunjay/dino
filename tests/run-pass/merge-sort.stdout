Enter some text: 
abcd

Enter some text: 
1234

Enter some text: 
abcd

Enter some text: 
abcd

Enter some text: 


Enter some text: 
1

Enter some text: 
111

Enter some text: 
0444557899

Enter some text: 
#$$())*2445668ABCDFGHLPPWXZZZ^^aabcd

Enter some text: 
!!!###$%&&&(())**0011112222344456677778899@@@AABCCCDDDEFFGGHHHHJJKKKKKKLMMMNNNNNNOPPPPQQQQRSSSSTTTUUVWWXZZ^^^^^aaaabbbcccccddfff

Enter some text: 
337DELMOPSVVWWXabbcde

Enter some text: 
AFLPTVWWYZ

Enter some text: 
aabccdddddefggghiiijkkkllllmmmmmmmnoopppppqqqqrrsssstwwwyyyzzz

Enter some text: 
